baxter_collaboration_lib/src/robot_interface/gripper.cpp

```diff
@@ -146,17 +203,279 @@ bool Gripper::has_error()
+        calibrate(true, 2.0);
+        ROS_INFO("Calibration complete");
+    }
+
+    // ensures that the gripper is positioned within physical limits
+    if(_position >= 0.0 && _position <= 100.0)
+    {
+        std::string position_cmd = EndEffectorCommand::CMD_GO;
+        std::string position_args = "{\"position\": " +
+                                     std::to_string(_position) + "}";
+
+        return command(position_cmd, _block, _timeout, position_args);
+    }
+    else
+    {
+        ROS_WARN("%s gripper of type %s must have position between 0.0 and 100.0",
+                  getGripperLimb().c_str(), type().c_str());
+        return false;
+    }
+}
+
+bool Gripper::commandSuction(bool _block, double _timeout)
+{
+    // give a warning if not capable and return
+    if(type() != "suction")
+    {
+        capabilityWarning("commandSuction");
+        return false;
+    }
+
+    std::string suction_cmd = EndEffectorCommand::CMD_GO;
+    std::string suction_args = "{\"grip_attempt_seconds\": " +
+                                std::to_string(_timeout) + "}";
+
+    return command(suction_cmd, _block, _timeout, suction_args);
+}
+
+bool Gripper::stop(bool _block, double _timeout)
+{
+    std::string stop_cmd;
+
+    if(type() == "electric")
+    {
+        // calibrate the electric gripper if needed
+        // do it in a blocking manner so that the gripper does not move before calibration
+        if(not is_calibrated())
+        {
+            ROS_INFO("Calibrating gripper. Please wait...");
+            calibrate(true, 2.0);
+            ROS_INFO("Calibration complete");
+        }
+
+        stop_cmd = EndEffectorCommand::CMD_STOP;
+    }
+    else if(type() == "suction")
+    {
+        stop_cmd = EndEffectorCommand::CMD_RELEASE;
+    }
+    else
+    {
+        // give a warning if not capable and return
+        capabilityWarning("stop");
+        return false;
+    }
+
+    command(stop_cmd, _block, _timeout);
+
     return true;
-    // return getGripperState().gripping==EndEffectorState::STATE_TRUE;
 }
 
-Gripper::~Gripper()
+bool Gripper::command(std::string _cmd, bool _block,
+                      double _timeout, std::string _args)
+{
+    EndEffectorCommand ee_cmd;
+    ee_cmd.id = get_id();
+    ee_cmd.command = _cmd;
+    ee_cmd.sender = cmd_sender;
+    ee_cmd.sequence = incCmdSeq();
+    ee_cmd.args = "";
+    if(_args != "")
+    {
+        ee_cmd.args = _args;
+    }
+
+    pub.publish(ee_cmd);
+
+    if(_block)
+    {
+        ros::Duration timeout(_timeout);
+        wait(timeout);
+    }
+
+    return true;
+}
+
+void Gripper::capabilityWarning(std::string _function)
+{
+    ROS_WARN("%s gripper of type %s is not capable of %s",
+              getGripperLimb().c_str(), type().c_str(), _function.c_str());
+}
+
+int Gripper::incCmdSeq()
+{
+    // prevents cmd_sequence from overflowing the integer limit
+    cmd_sequence = (cmd_sequence % std::numeric_limits<int>::max()) + 1;
```

Comment:
`incCmdSeq`? It increments the command sequence counter (wrapping it around the max int possible to prevent overflow). It is made use of by the `command()` method, and makes sure that the published messages to the gripper have an incremented sequence